language-plutus-core/src/Language/PlutusCore/TypeRenamer.hs

```diff
@@ -144,7 +144,7 @@ renameTerm st t@(LamAbs x (Name x' s (Unique u)) ty t') = do
         Just _ ->
             modify (+1) >>
             LamAbs x (Name x' s (Unique (m+1))) <$> renameType st' ty <*> renameTerm st' t'
-        _      -> pure t
+        _      -> renameTerm st t
```

Comment:
I am pretty sure, we need to use `st'` here. Otherwise, we'll fail to notice if an enclosed term use the same name as the one we just matched on.